File: java/src/main/java/com/entity/FenwickTree.java
package com.entity;

/**
 * @author: John
 * @date: 2020-03-23 18:30
 * @description: FenwickTree
 **/
public class FenwickTree {
    int sums_[];
    public FenwickTree(int n) {
       sums_ = new int[n + 1];
    }
    public void update(int i, int delta) {
        while (i < sums_.length) {
            sums_[i] += delta;
            i += i & -i;
        }
    }

    public int query(int i) {
        int sum = 0;
        while (i > 0) {
           sum += sums_[i];
           i -= i & -i;
        }
        return sum;
    }

}

File: leetcodepython/app/algorithm/FenwickTree.py
class FenwickTree:
    def __int__(self, n):
        self._sums = [0 for _ in range(n + 1)]

    def update(self, i, delta):
        while i < len(self._sums):
            self._sums[i] += delta
            i += i & -i

    def query(self, i):
        s = 0
        while i > 0:
            s += self._sums[i]
            i -= i & -i
        return s